### PR TITLE
feat: validate payment id size

### LINKED ISSUE #7096: ensure safe Payment id size
Ensure that the payment id of user data is limited. 
We use payment id to transmit wallet data to the sender as well as the user supplied payment id. 
We need to ensure that the user data is limited such that the wallet data will always be able to fit.

### PR DESCRIPTION
**Pull request by MCozhusheck**
Description
---
Closes: #7096 and #7052

Renames `PaymentId` to `MemoField` and adds constructor methods with input validation, as the total size of a `MemoField` is limited to 256 bytes.

Breaking Changes
---

- [ ] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [X] Other - Please specify

The `PaymentId` type has been renamed to `MemoField`. All external applications using this type will need to update their code to reflect the new naming.

## Summary by CodeRabbit

* **New Features**
  * Introduced strict validation and a 256-byte size limit for payment IDs, with clear error messages for invalid entries.
  * Added new methods for creating and inspecting payment IDs, improving flexibility and user feedback.
  * Added new error variants to improve reporting of invalid payment ID formats in transaction and output management services.

* **Bug Fixes**
  * Improved error handling and user feedback when entering invalid or oversized payment IDs in transaction forms.

* **Refactor**
  * Standardized payment ID creation and usage throughout the application for consistency and maintainability.
  * Simplified extraction of transaction and address details from payment IDs using new accessor methods.
  * Improved code clarity by replacing explicit pattern matching with method calls on payment IDs.
  * Reformatted code for better readability and removed unused imports.
  * Renamed payment ID fields and methods to "memo field" for clearer terminology.
  * Updated transaction, wallet, and output management modules to use the new memo field abstraction.

* **Tests**
  * Enhanced and expanded unit tests to cover new validation logic and edge cases for payment IDs.
  * Updated test cases to use new payment ID constructors and error handling methods.

### CODE REVIEW

**Comment by MCozhusheck** [comment]
Marking it as ready for review just to get some code rabbit comments, do not merge yet

File: base_layer/core/src/transactions/transaction_components/payment_id.rs [hunk 1]
```diff
@@ -39,6 +39,9 @@ use crate::transactions::{
     transaction_components::encrypted_data::{SIZE_U256, SIZE_VALUE},
 };
 
+// Maximum size for a PaymentID in bytes (256 bytes)
+const MAX_PAYMENT_ID_SIZE: usize = 256;
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/payment_id.rs]: this is a consensus constant

File: base_layer/core/src/transactions/transaction_components/payment_id.rs [hunk 2]
```diff
@@ -296,35 +501,39 @@ impl PaymentId {
...
-                sender_address,
-                sender_one_sided,
-                fee,
-                tx_type,
-                user_data,
+        match self.inner {
+            InnerPaymentId::Open { user_data, tx_type } => {
+                match PaymentId::new_address_and_data(sender_address, fee, sender_one_sided, tx_type, user_data) {
+                    Ok(payment_id) => payment_id,
+                    Err(e) => panic!("Cannot create AddressAndData PaymentId: {}", e),
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/payment_id.rs]: never panic, should create an error here, and propagate that back out.

File: base_layer/core/src/transactions/transaction_components/payment_id.rs [hunk 3]
```diff
@@ -121,7 +124,20 @@ impl Display for TxType {
 }
 
 #[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, Default)]
-pub enum PaymentId {
+pub struct PaymentId {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/payment_id.rs]:
```suggestion
pub struct MemoField {
```

File: base_layer/core/src/transactions/transaction_components/payment_id.rs [hunk 4]
```diff
@@ -121,7 +124,20 @@ impl Display for TxType {
...
+impl Deref for PaymentId {
+    type Target = InnerPaymentId;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, Default)]
+pub enum InnerPaymentId {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/payment_id.rs]:
```suggestion
enum InnerMemoField {
```
Dont expose this at all

**Review by SWvheerden — COMMENTED** [review]
This looking good and in the right direction. 
I do think while we are changing everything here, lets add this: https://github.com/tari-project/tari/issues/7052

Make sure we handle the errors properly and don't panic

File: base_layer/core/src/transactions/transaction_components/memo_field.rs [hunk 5]
```diff
@@ -0,0 +1,2331 @@
...
+        match &self.inner {
+            InnerMemoField::Open { tx_type, .. } |
+            InnerMemoField::AddressAndData { tx_type, .. } |
+            InnerMemoField::TransactionInfo { tx_type, .. } => *tx_type,
+            _ => TxType::default(),
+        }
+    }
+
+    /// Helper function to set the 'recipient_address' of a 'PaymentId::TransactionInfo'
+    pub fn transaction_info_set_address(&mut self, address: TariAddress) {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/memo_field.rs]: this can make the tx larger, this needs to check the size

File: base_layer/core/src/transactions/transaction_components/memo_field.rs [hunk 6]
```diff
@@ -0,0 +1,2331 @@
...
+        if let InnerMemoField::TransactionInfo {
+            ref mut recipient_address,
+            ..
+        } = self.inner
+        {
+            *recipient_address = address
+        }
+    }
+
+    pub fn transaction_info_set_sent_output_hashes(&mut self, sent_output_hashes: Vec<FixedHash>) {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/memo_field.rs]: same here, this adds 32 bytes for each hash

**Review by SWvheerden — COMMENTED** [review]
Looking very good, just two unchecked functions we need to ensure are fine

**Review by SWvheerden — APPROVED** [review]
